finetuner/tuner/dataset/samplers.py

```diff
@@ -63,12 +63,20 @@ def __init__(
         for key, val in self._class_to_labels.items():
             self._cls_group_counts.append([key, ceil(len(val) / num_items_per_class)])
 
-        # Approximate (max) length
-        self._length = ceil(
-            sum([x[1] for x in self._cls_group_counts]) / self._num_classes
-        )
+        # Prepare the current batches
+        self._prepare_batches()
 
     def __iter__(self) -> Iterator[Tuple[int, ...]]:
+        batches = self._batches
```

Comment:
this code seems weird.

Why do I get `batches` in advance before a call that changes `self._batches` to return the old ones?

Reply:
If you call `__len__` before calling `__iter__` you will get the wrong length if `self._batches` are not precomputed. So the only solution is to pre-compute the next `_batches` when you call `__iter__`. Because of random sampling `__len__` can not easily be computed

Reply:
I have now reformated this, so that the sampler is also the iterator, and new batches are only computed once the current ones are exhausted.